### PR TITLE
Dev/issue 7735

### PR DESCRIPTION
**Pull request by hammzj**
## Proposed changes

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)
* Fix for https://github.com/webdriverio/webdriverio/issues/7735
### Fix for [Issue 7735](https://github.com/webdriverio/webdriverio/issues/7735)

Updated `getPickleIds` to allow setting line numbers on Cucumber features. This method would pass in the absolute URI of the feature into the `pickleFilter.matches` method; the `matches` method expects a **relative** URI, so it will skip checking for line numbers if an absolute URI is passed in.

I did not add any new tests for it as there are relatively little within the `cucumberEventListener.test.ts` file. However, I provided my validations on the [Cucumber feature example](https://github.com/webdriverio/webdriverio/blob/main/examples/wdio/cucumber/features/my-feature.feature) below:

<details>
<summary>Normal test</summary>

```
$ npm run test:cucumber

> test:cucumber
> ../../packages/wdio-cli/bin/wdio.js ./cucumber/wdio.conf.js


Execution of 1 workers started at 2023-07-21T17:32:20.451Z

[0-0] RUNNING in chrome - /cucumber/features/my-feature.feature
[0-0] PASSED in chrome - /cucumber/features/my-feature.feature

 "spec" Reporter:
------------------------------------------------------------------
[Chrome 115.0.0.0 darwin #0-0] Running: Chrome (v115.0.0.0) on darwin
[Chrome 115.0.0.0 darwin #0-0] Session ID: 80c074fb-4015-4209-90b5-58ce8c0a02cd
[Chrome 115.0.0.0 darwin #0-0]
[Chrome 115.0.0.0 darwin #0-0] » /cucumber/features/my-feature.feature
[Chrome 115.0.0.0 darwin #0-0] Example feature
[Chrome 115.0.0.0 darwin #0-0] As a user of WebdriverIO
[Chrome 115.0.0.0 darwin #0-0] I should be able to use different commands
[Chrome 115.0.0.0 darwin #0-0] to get informations about elements on the page
[Chrome 115.0.0.0 darwin #0-0]
[Chrome 115.0.0.0 darwin #0-0] Get size of an element
[Chrome 115.0.0.0 darwin #0-0]    ✓ Given I go on the website "https://github.com/"
[Chrome 115.0.0.0 darwin #0-0]    ✓ Then should the element ".header-logged-out a" be 32px wide and 34.5px high
[Chrome 115.0.0.0 darwin #0-0]
[Chrome 115.0.0.0 darwin #0-0] Get title of website
[Chrome 115.0.0.0 darwin #0-0] Business rule 1
[Chrome 115.0.0.0 darwin #0-0]    ✓ Given I go on the website "https://github.com/"
[Chrome 115.0.0.0 darwin #0-0]    ✓ Then should the title of the page be "GitHub: Let’s build from here · GitHub"
[Chrome 115.0.0.0 darwin #0-0]
[Chrome 115.0.0.0 darwin #0-0] Data Tables
[Chrome 115.0.0.0 darwin #0-0] Business rule 2
[Chrome 115.0.0.0 darwin #0-0]    ✓ Given I go on the website "http://todomvc.com/examples/react/#/"
[Chrome 115.0.0.0 darwin #0-0]    ✓ When I add the following groceries
[Chrome 115.0.0.0 darwin #0-0]      │ Item       │ Amount │
[Chrome 115.0.0.0 darwin #0-0]      │ Milk       │ 2      │
[Chrome 115.0.0.0 darwin #0-0]      │ Butter     │ 1      │
[Chrome 115.0.0.0 darwin #0-0]      │ Noodles    │ 1      │
[Chrome 115.0.0.0 darwin #0-0]      │ Schocolate │ 3      │
[Chrome 115.0.0.0 darwin #0-0]    ✓ Then I should have a list of 4 items
[Chrome 115.0.0.0 darwin #0-0]
[Chrome 115.0.0.0 darwin #0-0] 7 passing (4.1s)


Spec Files:      1 passed, 1 total (100% completed) in 00:00:11 
```
</details>

<details>
<summary>Test with `--spec` option</summary>

```
$ npm run test:cucumber -- --spec ./cucumber/features/my-feature.feature

> test:cucumber
> ../../packages/wdio-cli/bin/wdio.js ./cucumber/wdio.conf.js --spec ./cucumber/features/my-feature.feature


Execution of 1 workers started at 2023-07-21T17:34:18.530Z

[0-0] RUNNING in chrome - /cucumber/features/my-feature.feature
[0-0] PASSED in chrome - /cucumber/features/my-feature.feature

 "spec" Reporter:
------------------------------------------------------------------
[Chrome 115.0.0.0 darwin #0-0] Running: Chrome (v115.0.0.0) on darwin
[Chrome 115.0.0.0 darwin #0-0] Session ID: 4ba0b68e-4e28-4de7-bbd1-00d3993eba2b
[Chrome 115.0.0.0 darwin #0-0]
[Chrome 115.0.0.0 darwin #0-0] » /cucumber/features/my-feature.feature
[Chrome 115.0.0.0 darwin #0-0] Example feature
[Chrome 115.0.0.0 darwin #0-0] As a user of WebdriverIO
[Chrome 115.0.0.0 darwin #0-0] I should be able to use different commands
[Chrome 115.0.0.0 darwin #0-0] to get informations about elements on the page
[Chrome 115.0.0.0 darwin #0-0]
[Chrome 115.0.0.0 darwin #0-0] Get size of an element
[Chrome 115.0.0.0 darwin #0-0]    ✓ Given I go on the website "https://github.com/"
[Chrome 115.0.0.0 darwin #0-0]    ✓ Then should the element ".header-logged-out a" be 32px wide and 34.5px high
[Chrome 115.0.0.0 darwin #0-0]
[Chrome 115.0.0.0 darwin #0-0] Get title of website
[Chrome 115.0.0.0 darwin #0-0] Business rule 1
[Chrome 115.0.0.0 darwin #0-0]    ✓ Given I go on the website "https://github.com/"
[Chrome 115.0.0.0 darwin #0-0]    ✓ Then should the title of the page be "GitHub: Let’s build from here · GitHub"
[Chrome 115.0.0.0 darwin #0-0]
[Chrome 115.0.0.0 darwin #0-0] Data Tables
[Chrome 115.0.0.0 darwin #0-0] Business rule 2
[Chrome 115.0.0.0 darwin #0-0]    ✓ Given I go on the website "http://todomvc.com/examples/react/#/"
[Chrome 115.0.0.0 darwin #0-0]    ✓ When I add the following groceries
[Chrome 115.0.0.0 darwin #0-0]      │ Item       │ Amount │
[Chrome 115.0.0.0 darwin #0-0]      │ Milk       │ 2      │
[Chrome 115.0.0.0 darwin #0-0]      │ Butter     │ 1      │
[Chrome 115.0.0.0 darwin #0-0]      │ Noodles    │ 1      │
[Chrome 115.0.0.0 darwin #0-0]      │ Schocolate │ 3      │
[Chrome 115.0.0.0 darwin #0-0]    ✓ Then I should have a list of 4 items
[Chrome 115.0.0.0 darwin #0-0]
[Chrome 115.0.0.0 darwin #0-0] 7 passing (3.7s)


Spec Files:      1 passed, 1 total (100% completed) in 00:00:09 

```
</details>

<details>
<summary>Test with single line number</summary>

```
$ npm run test:cucumber -- --spec ./cucumber/features/my-feature.feature:11

> test:cucumber
> ../../packages/wdio-cli/bin/wdio.js ./cucumber/wdio.conf.js --spec ./cucumber/features/my-feature.feature:11


Execution of 1 workers started at 2023-07-21T17:33:16.147Z

[0-0] RUNNING in chrome - /cucumber/features/my-feature.feature
[0-0] PASSED in chrome - /cucumber/features/my-feature.feature

 "spec" Reporter:
------------------------------------------------------------------
[Chrome 115.0.0.0 darwin #0-0] Running: Chrome (v115.0.0.0) on darwin
[Chrome 115.0.0.0 darwin #0-0] Session ID: 01952991-33c2-4094-a80b-ffdcce7305d6
[Chrome 115.0.0.0 darwin #0-0]
[Chrome 115.0.0.0 darwin #0-0] » /cucumber/features/my-feature.feature
[Chrome 115.0.0.0 darwin #0-0] Example feature
[Chrome 115.0.0.0 darwin #0-0] As a user of WebdriverIO
[Chrome 115.0.0.0 darwin #0-0] I should be able to use different commands
[Chrome 115.0.0.0 darwin #0-0] to get informations about elements on the page
[Chrome 115.0.0.0 darwin #0-0]
[Chrome 115.0.0.0 darwin #0-0] Get title of website
[Chrome 115.0.0.0 darwin #0-0] Business rule 1
[Chrome 115.0.0.0 darwin #0-0]    ✓ Given I go on the website "https://github.com/"
[Chrome 115.0.0.0 darwin #0-0]    ✓ Then should the title of the page be "GitHub: Let’s build from here · GitHub"
[Chrome 115.0.0.0 darwin #0-0]
[Chrome 115.0.0.0 darwin #0-0] 2 passing (1.9s)


Spec Files:      1 passed, 1 total (100% completed) in 00:00:08 
```
</details>

<details>
<summary>Test with `--spec` and multiple line numbers</summary>

```
$ npm run test:cucumber -- --spec ./cucumber/features/my-feature.feature:11:16

> test:cucumber
> ../../packages/wdio-cli/bin/wdio.js ./cucumber/wdio.conf.js --spec ./cucumber/features/my-feature.feature:11:16


Execution of 1 workers started at 2023-07-21T17:36:15.811Z

[0-0] RUNNING in chrome - /cucumber/features/my-feature.feature
[0-0] PASSED in chrome - /cucumber/features/my-feature.feature

 "spec" Reporter:
------------------------------------------------------------------
[Chrome 115.0.0.0 darwin #0-0] Running: Chrome (v115.0.0.0) on darwin
[Chrome 115.0.0.0 darwin #0-0] Session ID: 23d9ca30-07ec-437e-8131-9f97a79bf543
[Chrome 115.0.0.0 darwin #0-0]
[Chrome 115.0.0.0 darwin #0-0] » /cucumber/features/my-feature.feature
[Chrome 115.0.0.0 darwin #0-0] Example feature
[Chrome 115.0.0.0 darwin #0-0] As a user of WebdriverIO
[Chrome 115.0.0.0 darwin #0-0] I should be able to use different commands
[Chrome 115.0.0.0 darwin #0-0] to get informations about elements on the page
[Chrome 115.0.0.0 darwin #0-0]
[Chrome 115.0.0.0 darwin #0-0] Get title of website
[Chrome 115.0.0.0 darwin #0-0] Business rule 1
[Chrome 115.0.0.0 darwin #0-0]    ✓ Given I go on the website "https://github.com/"
[Chrome 115.0.0.0 darwin #0-0]    ✓ Then should the title of the page be "GitHub: Let’s build from here · GitHub"
[Chrome 115.0.0.0 darwin #0-0]
[Chrome 115.0.0.0 darwin #0-0] Data Tables
[Chrome 115.0.0.0 darwin #0-0] Business rule 2
[Chrome 115.0.0.0 darwin #0-0]    ✓ Given I go on the website "http://todomvc.com/examples/react/#/"
[Chrome 115.0.0.0 darwin #0-0]    ✓ When I add the following groceries
[Chrome 115.0.0.0 darwin #0-0]      │ Item       │ Amount │
[Chrome 115.0.0.0 darwin #0-0]      │ Milk       │ 2      │
[Chrome 115.0.0.0 darwin #0-0]      │ Butter     │ 1      │
[Chrome 115.0.0.0 darwin #0-0]      │ Noodles    │ 1      │
[Chrome 115.0.0.0 darwin #0-0]      │ Schocolate │ 3      │
[Chrome 115.0.0.0 darwin #0-0]    ✓ Then I should have a list of 4 items
[Chrome 115.0.0.0 darwin #0-0]
[Chrome 115.0.0.0 darwin #0-0] 5 passing (3s)


Spec Files:      1 passed, 1 total (100% completed) in 00:00:08 

```
</details>

The only issue I faced is that I could not run more than two scenarios at a time this way:
```
# Does not work!
$ npm run test:cucumber -- --spec ./cucumber/features/my-feature.feature:6:11:16
```

 This [StackOverflow](https://stackoverflow.com/a/32141981) answer said it should be possible, but this might be a Cucumber-js issue and not a WDIO issue.

***

### Small updates to Cucumber tests
GitHub's website changed, so I updated the `my-feature.feature` test document. These test are incredibly flaky since they rely on an third-party website to validate. It would make more sense to validate against  a WebDriverIO test website or homepage where the maintainers know what changes are made to their website.

As well, these steps should move away from regular expressions in favor of Cucumber Expressions. [See here](https://cucumber.io/docs/cucumber/cucumber-expressions/).

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] I have added proper type definitions for new commands (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

File: packages/wdio-cucumber-framework/src/cucumberEventListener.ts [hunk 1]
```diff
@@ -13,10 +13,12 @@ import type { Capabilities } from '@wdio/types'
 import { addKeywordToStep, filterPickles, getRule } from './utils.js'
 import type { ReporterScenario } from './constants.js'
 import type { HookParams } from './types.js'
+import path from 'path'
 
 const log = logger('CucumberEventListener')
 
 export default class CucumberEventListener extends EventEmitter {
+    private _cwd: string
```
Comment (christian-bromann) [inline packages/wdio-cucumber-framework/src/cucumberEventListener.ts:21]:
```suggestion
    #cwd = process.cwd()
```

File: packages/wdio-cucumber-framework/src/cucumberEventListener.ts [hunk 2]
```diff
@@ -29,6 +31,7 @@ export default class CucumberEventListener extends EventEmitter {
 
     constructor (eventBroadcaster: EventEmitter, private _pickleFilter: InstanceType<typeof PickleFilter>) {
         super()
+        this._cwd = process.cwd()
```
Comment (christian-bromann) [inline packages/wdio-cucumber-framework/src/cucumberEventListener.ts:34]:
```suggestion
```

File: packages/wdio-cucumber-framework/src/cucumberEventListener.ts [hunk 3]
```diff
@@ -441,10 +444,14 @@ export default class CucumberEventListener extends EventEmitter {
             /**
              * match based on Cucumber pickle filter
              */
-            .filter(([, fakeId]) => this._pickleFilter.matches({
-                gherkinDocument,
-                pickle: this._scenarios.find(s => s.id === fakeId) as Pickle
-            }))
+            .filter(([, fakeId]) => {
+                const pickle = { ...this._scenarios.find(s => s.id === fakeId) } as Pickle
+                pickle.uri = path.relative(this._cwd, pickle.uri)
```
Comment (christian-bromann) [inline packages/wdio-cucumber-framework/src/cucumberEventListener.ts:449]:
```suggestion
                pickle.uri = path.relative(this.#cwd, pickle.uri)
```

File: examples/wdio/cucumber/step-definitions.js [hunk 4]
```diff
@@ -28,11 +28,11 @@ When(/^I add the following groceries$/, async (table) => {
     }
 })
 
-Then(/^should the element "([^"]*)" be (\d+)px wide and (\d+)px high$/, async (selector, width, height) => {
-    const elemSize = await $(selector).getSize()
-    expect(elemSize.width).toBe(width)
-    expect(elemSize.height).toBe(height)
-})
+Then(/^should the element "([^"]*)" be (\d+(?:\.\d+)?)px wide and (\d+(?:\.\d+)?)px high$/, async (selector, width, height) => {
+    var elemSize = await $(selector).getSize()
```
Comment (christian-bromann) [inline examples/wdio/cucumber/step-definitions.js:32]:
```suggestion
    const elemSize = await $(selector).getSize()
```

**Review by christian-bromann — COMMENTED** [review]
I wonder if it makes sense to add a smoke test for this to make sure it works.

File: packages/wdio-cucumber-framework/src/cucumberEventListener.ts (hunk 1 above)
Comment (hammzj) [reply packages/wdio-cucumber-framework/src/cucumberEventListener.ts:21]: Why remove `cwd` here? We need to pass it in to get the relative path of the pickle URI.

Comment (christian-bromann) [reply packages/wdio-cucumber-framework/src/cucumberEventListener.ts:21]: `cwd` is not removed, I would just prefer to define the variable as JS private rather then TS private variable, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields

Comment (hammzj) [reply packages/wdio-cucumber-framework/src/cucumberEventListener.ts]: Never knew this was a thing. awesome

**Comment by hammzj** [comment]
I'll look at adding smoke tests next week. It's Friday, i gotta hit the gym and then start a weekend-long happy hour

**Review by christian-bromann — APPROVED** [review]
Great changeset .. awesome job 👍 
Have a great weekend!

**Pull request merged**